src/Microsoft.AspNetCore.Blazor.Browser.JS/src/Interop/DotNetInvoker.ts

```diff
@@ -0,0 +1,103 @@
+﻿import { MethodOptions, DotnetMethodArgumentsList } from './MethodOptions';
+import { platform } from '../Environment';
+import { System_String } from '../Platform/Platform';
+import { getRegisteredFunction } from './RegisteredFunction';
+
+export function invokeDotNetMethod<T>(methodOptions: MethodOptions, ...args: any[]): (T | null) {
+    const method = platform.findMethod(
+        "Microsoft.AspNetCore.Blazor.Browser",
+        "Microsoft.AspNetCore.Blazor.Browser.Interop",
+        "JavaScriptInvoke",
+        "InvokeDotnetMethod");
+
+    const packedArgs = packArguments(args);
+
+    const serializedOptions = platform.toDotNetString(JSON.stringify(methodOptions));
+    const serializedArgs = platform.toDotNetString(JSON.stringify(packedArgs));
+    const serializedResult = platform.callMethod(method, null, [serializedOptions, serializedArgs]);
+
+    if (serializedResult !== null && serializedResult !== undefined && (serializedResult as any) !== 0) {
+        const result = JSON.parse(platform.toJavaScriptString(serializedResult as System_String));
+        return result;
+    }
+
+    return null;
+}
+
+function packArguments(args: any[]): DotnetMethodArgumentsList {
+    var result = {};
+    if (args.length == 0) {
+        return result;
+    }
+
+    if (args.length > 7) {
+        for (let i = 0; i < 7; i++) {
+            result[`argument${[i + 1]}`] = args[i];
+        }
+        result['argument8'] = packArguments(args.slice(7));
+    } else {
+        for (let i = 0; i < args.length; i++) {
+            result[`argument${[i + 1]}`] = args[i];
+        }
+    }
+
+    return result;
+}
+
+let globalId = 0;
+
+export function invokeDotNetMethodAsync<T>(methodOptions: MethodOptions, ...args: any[]): Promise<T | null> {
+    const resolveId = (globalId++).toString();
+    const rejectId = (globalId++).toString();
+    methodOptions.async = { resolveId: resolveId, rejectId: rejectId, functionName: "invokeJavaScriptCallback" };
+
+    const result = new Promise<T | null>((resolve, reject) => {
+        TrackedReference.track(resolveId, resolve);
+        TrackedReference.track(rejectId, reject);
+    });
+
+    invokeDotNetMethod(methodOptions, ...args);
+
+    return result;
+}
+
+export function invokeJavaScriptCallback(id: string, ...args: any[]): void {
+    const callbackRef = TrackedReference.get(id);
+    const callback = callbackRef.trackedObject as Function;
+    callback.apply(null, args);
+}
+
+class TrackedReference {
+    private static references: Map<string, any> = new Map<string, any>();
+
+    private constructor(public id: string, public trackedObject: any) {
+    }
+
+    public static track(id: string, trackedObject: any): void {
```

Comment:
It's a minor suggestion, but instead of having to check whether the `id` values are already in use, you could have the `track` function responsible for choosing the `id` and use that as the return value.